### PR TITLE
HPCC-26422 Initial Dali metrics

### PR DESCRIPTION
**Pull request by kenrowland**
Added some initial metrics to Dali

Signed-off-by: Ken Rowland kenneth.rowland@lexisnexisrisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by jakesmith — COMMENTED** [review]
@kenrowland - please don't squash until review is over, it makes it harder to see which review comments have been address/what has changed.

it looks fine I think, but should remove the redundant code in daserver.cpp lines 778-781, and make the ScopedGaugeUpdater usage consistent.

File: dali/base/danqs.cpp
Comment (ghalliday) [reply dali/base/danqs.cpp:45]: It should either be hpcc.nq_requests or dali.nq_requests.  Probably slight preference for the latter.

@jake would you commonly want to aggregate nq and sds requests together or not?

Comment (ghalliday) [reply dali/base/danqs.cpp:45]: A bit verbose, I wonder whether 
    hpccMetrics::registerCounterMetric(...)
would be better to improve readability.

File: dali/base/danqs.cpp
Comment (ghalliday) [reply dali/base/danqs.cpp:563]: This isn't very useful.  Its value will be either 0 or 1, indicating whether dali is currently processing a message.

File: dali/base/danqs.cpp
Comment (ghalliday) [inline dali/base/danqs.cpp:537]: @jakesmith I think this is in the wrong place - this is before the message has been received, so will generally be too high.

File: dali/base/danqs.cpp
Comment (ghalliday) [reply dali/base/danqs.cpp:563]: A general efficiency concern - this involves two atomic operations.  If you recorded the "dali.nq_requests.processed" instead it would only involve 1 atomic operation.  You could publish dali.nq_requests.active by subtracting them.  I suspect it is more useful to publish a counter of the number processed.
@rpastrana how easy is it in prometheus to subtract two counters and display the result?

Comment (ghalliday) [reply dali/base/danqs.cpp:563]: A different efficiency concern is that this code is single-threaded so none of the metric increments need to be atomic.  There should be a version of the increment that avoids the atomic penalty.  (approx 10ns on intel for reference, different on ARM/POWER8)

File: dali/base/dasds.cpp
Comment (ghalliday) [reply dali/base/dasds.cpp:98]: Number of SDS requests received, rather than handled.  Also needs a "dali." prefix.

File: dali/base/dasds.cpp
Comment (ghalliday) [inline dali/base/dasds.cpp:100]: I'm not sure how useful this metrics are for alerting or scaling.  I am happy for them to be published - because they will probably look pretty in a graph, but they're only a snapshot.  Publishing the counts allows you to work out the average load over the period, rather than the instantaneous load at the point the metrics were gathered.)

Better to publish "dai.sqs_requests.started", "dai.sqs_requests.processed" or "dai.sqs_requests.completed" (should be consistent with nq above, and same pattern should be used for all types of requests processed in the system.

File: dali/base/dasds.cpp
Comment (ghalliday) [inline dali/base/dasds.cpp:3763]: This doesn't need to be atomic.  It also replaces msgCount, so that code can be removed.

File: dali/base/dasds.cpp
Comment (ghalliday) [reply dali/base/dasds.cpp:4110]: This can also be non atomic.

If you only use counters I think all 3 counters can currently be non-atomic.

**Review by ghalliday — COMMENTED** [review]
@kenrowland some initial comments

**Comment by kenrowland** [comment]
@ghalliday @jakesmith Please review. A couple of things:
- I left some old code commented for reference to what was being done before
- I added a scoped increment of a count for the synthesized pending gauge (in order to ensure the number of completed requests is updated
- I did not change metric updates to the fast versions. I figure we can change them once we agree fully on the new names and which updates can be the fast versions (probably all if I remember correctly).

File: dali/server/daserver.cpp
Comment (ghalliday) [inline dali/server/daserver.cpp:788]: I don't see a matching call to stopCollecting()

Comment (ghalliday) [reply dali/server/daserver.cpp:788]: Thinking about generalizing this to more components.  Is it worth adding an 

initAndStartCollecting(IPropertyTree * optTree);

function so that instead of this if() block being duplicated in each component it could say

hpccMetrics::queryMetricsManager().initAndStartCollecting(pMetricsTree);

Also, since the code code for getting the pMetricsTree is likely to be identical for each component, even better might be:

hpccMetrics::queryMetricsManager().initAndStartCollecting();

File: esp/platform/txsummary.cpp
Comment (ghalliday) [inline esp/platform/txsummary.cpp:27]: description could be more verbose.  Same for others that have been modified.

File: system/jlib/jmetrics.hpp
Comment (ghalliday) [inline system/jlib/jmetrics.hpp:406]: is this useful?  I think it is implied by copy constructor on the  line above.  Same for the second assignment.  Did you mean to use &&?  I'm not sure it is likely to catch real world problems.

**Review by ghalliday — COMMENTED** [review]
@kenrowland A few comments, but looks closer to being ready.

File: system/jlib/jmetrics.hpp
Comment (kenrowland) [reply system/jlib/jmetrics.hpp]: After some further research, the non const copy constructor and assignment operator should be sufficient. Move copy constructors and assignment are not automatically generated, so they were left out. However, since we are deleting the non move copy constructor, best to ensure they are deleted as well.

File: testing/unittests/metrics/MetricFrameworkTests.cpp
Comment (kenrowland) [reply testing/unittests/metrics/MetricFrameworkTests.cpp:93]: I found that as I added more updates to the metric, keeping track of what the expected result was harder. An earlier test in the sequence forces a change in all subsequent asserts. So I added a variable to keep track of the expected changes. If you feel that a constant is better, let me know and I will change it back.

File: testing/unittests/metrics/MetricFrameworkTests.cpp
Comment (kenrowland) [reply testing/unittests/metrics/MetricFrameworkTests.cpp]: ScopedCountIncrement has been removed in favor of using COnScopeExit

**Comment by kenrowland** [comment]
@ghalliday @jakesmith Please see latest changes. A couple of open issues for Jake to comment and some questions.

@afishbeck Some minor changes to the implementation in ESP should you want to take a look.

File: dali/base/danqs.cpp
Comment (jakesmith) [reply dali/base/danqs.cpp:536]: yes, comments are in the wrong place.
Comment+code should be after the recv, or better still to be inside and at start of processMessage

File: dali/base/dasds.cpp
Comment (jakesmith) [reply dali/base/dasds.cpp:3758]: yes, this is the only thread that receives requests/can update.

File: dali/base/danqs.cpp
Comment (jakesmith) [reply dali/base/danqs.cpp]: >@jake would you commonly want to aggregate nq and sds requests together or not?

I don't think so. They are different beasts.

Comment (jakesmith) [reply dali/base/danqs.cpp]: > A bit verbose, I wonder whether
> hpccMetrics::registerCounterMetric(...)
> would be better to improve readability.

I agree, think that is more readable.

File: dali/base/dasds.cpp
Comment (jakesmith) [inline dali/base/dasds.cpp:4096]: inc(1) vs inc() inconsistency.

Unless @ghalliday disagrees, let's take out the default param value altogether and force it to [consistently] require to be inc(1)

File: dali/base/danqs.cpp
Comment (kenrowland) [reply dali/base/danqs.cpp]: Moved to inside processMessage

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp]: Changed to fastInc;

File: dali/base/dasds.cpp
Comment (kenrowland) [inline dali/base/dasds.cpp:4096]: @jakesmith Could the inc here also be a fastInc ?

File: dali/base/dasds.cpp
Comment (jakesmith) [inline dali/base/dasds.cpp:3760]: > This doesn't need to be atomic. It also replaces msgCount, so that code can be removed.

all references to msgCount should now be removed (2), as replaced by pSdsRequestsReceived

**Comment by jakesmith** [comment]
>If you only use counters I think all 3 counters can currently be non-atomic.

comment from changed code in sds, but..

@ghalliday , the started/completed do need to atomic afaics. There are multiple 'processMessage' threads that will be updating these counters.

File: dali/base/danqs.cpp
Comment (jakesmith) [inline dali/base/danqs.cpp:43]: the above comment block re. active should be removed as we're not intending to supply this (as can only be 0 or 1) for NQ.

File: dali/base/dasds.cpp
Comment (jakesmith) [inline dali/base/dasds.cpp:101]: did we drop the idea of tracking pending request?
I think that makes sense, because we know # received, number started and number completed so can calculate number blocked pending.
But if so, this line should be deleted.

File: dali/base/dasds.cpp
Comment (jakesmith) [reply dali/base/dasds.cpp:4096]: no, I don't see how it can, there will multiple concurrent threads at this point (in processMessage via thread pool)

File: system/jlib/jmetrics.hpp
Comment (jakesmith) [inline system/jlib/jmetrics.hpp:160]: Isn't this actually slower ..?
Because it's still an atomic "value + val" will require 'value' to be read atomically, and then the assignment requires an atomic assignment .
So potentially twice as slow..

@ghalliday - should the fast version instead use memory_order_relaxed order? (but would it actually be any quicker, given 2 ops).

And/or for places where you know the counter is only every going to be single threaded,
should there be a different implementation, e.g. CounterNonAtomicMetric

File: system/jlib/jmetrics.hpp
Comment (jakesmith) [inline system/jlib/jmetrics.hpp:181]: as above, I think this it's actually slower.

File: system/jlib/jmetrics.hpp
Comment (ghalliday) [reply system/jlib/jmetrics.hpp:160]: @jakesmith this is actually faster on intel.   The reason is that read and write are actually always atomic (unless they are misaligned, I think over a cache line), and it has a naturally strong memory ordering.  An atomic increment uses a LOCK ADD - which has various effects - mainly synchronizing the cache line over different cores, and has an overhead of ~18cycles as a result (read https://www.agner.org/optimize/instruction_tables.pdf if you really care!).

On non intel chips with a weak memory model there is likely to be a difference between using memory_order_relaxed - which is what the RelaxedAtomic defaults to - because the default std::atomic mode will probably require extra work to ensure the results are consistent (which isn't needed here).

So yes, this is faster, but even better would be to use RelaxedAtomic, or use the memory_order_relaxed on the access of the metric values,

File: dali/base/dasds.cpp
Comment (ghalliday) [reply dali/base/dasds.cpp:101]: @jakesmith this line automatically calculates and publishes the number.

File: system/jlib/jmetrics.hpp
Comment (jakesmith) [reply system/jlib/jmetrics.hpp:160]: ok, it's less than obvious from a casual glance at the code.

File: dali/base/dasds.cpp
Comment (jakesmith) [reply dali/base/dasds.cpp:101]: right ok, yes I'd missed what it was doing.

**Comment by ghalliday** [comment]
@kenrowland after more thought and research I think we need to get the memory order parameter correct - and make sure it is relaxed.  (It does not matter if these atomic reads and writes are reordered in relation to other reads and writes.)

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp:4095]: Comment was informational only for the review in case anyone wanted to know what change this represented. I intended on removing it for the final commit. I will remove it now.

File: dali/base/danqs.cpp
Comment (kenrowland) [reply dali/base/danqs.cpp:43]: Removed.

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp:3756]: Removed

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp:3760]: Removed msgCount.

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp:4096]: Since that is change to the base framework, I would like to address it in a separate Jira after this is merged. @ghalliday, comments?

Comment (kenrowland) [reply dali/base/dasds.cpp:4096]: It shall remain as is.

File: esp/platform/espprotocol.cpp
Comment (kenrowland) [reply esp/platform/espprotocol.cpp:33]: Added.

File: system/jlib/jmetrics.hpp
Comment (kenrowland) [reply system/jlib/jmetrics.hpp:198]: Let's go with being and end for now.

File: system/jlib/jmetrics.hpp
Comment (kenrowland) [reply system/jlib/jmetrics.hpp:203]: Moved

**Comment by kenrowland** [comment]
@ghalliday In order to get the memory order correct per your comment, I went ahead and changed the type for the atomic member variable to a RelaxedAtomic instead of adding the order parameter to each individual operation.

File: dali/base/dasds.cpp
Comment (ghalliday) [inline dali/base/dasds.cpp:84]: Minor: The following comments are also redundant and should go.

File: dali/base/dasds.cpp
Comment (ghalliday) [reply dali/base/dasds.cpp:4089]: minor: not removed yet.

File: dali/base/dasds.cpp
Comment (ghalliday) [reply dali/base/dasds.cpp:4090]: I don't mind if this pr, or the next.  The next should also revisit all the existing metric names and make sure they match the agreed naming scheme.  (I think they may already, but worth double checking.)

File: dali/server/daserver.cpp
Comment (ghalliday) [inline dali/server/daserver.cpp:772]: @jakesmith would it be suitable to call stopCollecting() after this call to stopServer()?

File: system/jlib/jmetrics.hpp
Comment (ghalliday) [inline system/jlib/jmetrics.hpp:143]: remove comment

File: system/jlib/jmetrics.hpp
Comment (ghalliday) [reply system/jlib/jmetrics.hpp:162]: should now call value.fastAdd(val);

File: system/jlib/jmetrics.hpp
Comment (ghalliday) [inline system/jlib/jmetrics.hpp:184]: should now call value.fastAdd()

File: system/jlib/jmetrics.hpp
Comment (ghalliday) [inline system/jlib/jmetrics.hpp:211]: really picky (I debated even mentioning it).
Better as

```
endValue = pEndCounter->queryValue()
return pBeginCounter->queryValue() - endValue;

```
Why?  The evaluation order is undefined in the previous code.  If begin->value was called, then both begin and end were incremented before end->value() was called you could end up with a -ve count.  Making the order explicit will ensure it is positive.

**Review by ghalliday — COMMENTED** [review]
@kenrowland a few comments.  Looks like it is close to being ready.

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp:84]: In my opinion these comments are now no longer needed. The instantiation of the metric itself has the description which should suffice. If additional comments are needed referencing the individual metrics, I can restore them.

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp:4089]: I actually removed it this time.

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp:4090]: Opened https://track.hpccsystems.com/browse/HPCC-26501 to address this.

File: system/jlib/jmetrics.hpp
Comment (kenrowland) [reply system/jlib/jmetrics.hpp:143]: Removed (yes actually removed).

File: system/jlib/jmetrics.hpp
Comment (kenrowland) [reply system/jlib/jmetrics.hpp:162]: Updated to use relaxed atomic class methods.

File: system/jlib/jmetrics.hpp
Comment (kenrowland) [reply system/jlib/jmetrics.hpp:184]: Updated.

File: system/jlib/jmetrics.hpp
Comment (kenrowland) [reply system/jlib/jmetrics.hpp:211]: An explicit order is better. The value can be 0 (which some argue is included with a positive designation).

**Review by kenrowland — COMMENTED** [review]
Updated based on comments from @ghalliday.

@jakesmith There are a few open questions to you.

File: dali/base/dasds.cpp
Comment (jakesmith) [reply dali/base/dasds.cpp]: I don't think any of the remaining comments above useful (and slightly confusing)

File: dali/base/dasds.cpp
Comment (jakesmith) [reply dali/base/dasds.cpp:4078]: >Since that is change to the base framework, I would like to address it in a separate Jira after this is merged.

without chanigng the methods, can we at least make the call consistent? i.e. make both the above calls inc(1).

File: dali/server/daserver.cpp
Comment (jakesmith) [inline dali/server/daserver.cpp:769]: this isn't going to work here, covenMain enters a loop and will not return until Dali starts to quit.
Should be moved before covenMain call

Comment (kenrowland) [reply dali/server/daserver.cpp:769]: @jakesmith Can it simply be moved up one line?

File: dali/server/daserver.cpp
Comment (jakesmith) [reply dali/server/daserver.cpp:772]: yes, think it should be.
If it isn't I suspect there could be potential problem (crashes) during closedown whilst the sink threads are still running and objects they are reporting on are destructed.

Comment (kenrowland) [reply dali/server/daserver.cpp:772]: @jakesmith The framework only maintains weak pointers to metric objects, so if Dali is shutting down and destroying any metric objects, the framework will detect that the object has been deleted and will delete it from its internal collection.

Comment (ghalliday) [reply dali/server/daserver.cpp:772]: It can still cause a problem - if the framework has a strong pointer to an object, and that object is an instance of a class defined in a dll which is then unloaded.  Then the thread can crash when it calls a virtual method of the class whose definition is no longer in memory.  We have even seen instances of this in the past...

**Review by jakesmith — CHANGES_REQUESTED** [review]
@kenrowland - a few comments.
The significant one is that dali metrics can't be working at the moment afaics,
because initializeMetrics is only being called after Dali starts exiting.

File: dali/server/daserver.cpp
Comment (kenrowland) [reply dali/server/daserver.cpp:772]: That is a danger when using custom metrics and perhaps we should better document the concern. Or, custom metrics should be a temporary solution with an impetus to migrate to framework defined classes.

File: dali/server/daserver.cpp
Comment (jakesmith) [reply dali/server/daserver.cpp:768]: yes that is enough afaics.

File: dali/server/daserver.cpp
Comment (jakesmith) [reply dali/server/daserver.cpp]: If you add stopCollecting() call, the threads I assume will be stopped, therefore avoiding the problem being discussed.

File: testing/unittests/metrics/MetricFrameworkTests.cpp
Comment (ghalliday) [reply testing/unittests/metrics/MetricFrameworkTests.cpp:93]: I don't really understand this - if you add tests on the end then you don't need to change any other code, and the previously value is the one in the assert.
Not worth the hassle of updating though.

**Review by ghalliday — COMMENTED** [review]
@kenrowland a few notes about out of date comments, but looks very close to ready.  Should be squashed once others have approved.

File: dali/base/dasds.cpp
Comment (kenrowland) [reply dali/base/dasds.cpp]: Add comment to make it clear what is happening

**Comment by kenrowland** [comment]
@ghalliday Rebased with latest comment changes.
@jakesmith If you want a final review...

**Review by ghalliday — APPROVED** [review]

**Comment by kenrowland** [comment]
@richardkchapman Please merge

**Review by jakesmith — APPROVED** [review]

**Pull request merged**
